## ひろしまQuest2020#stayhome【球種予測部門】

In [23]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import lightgbm as lgb
import gc
%matplotlib inline
pd.set_option('display.max_Columns', 100)

In [24]:
ALL_MERGE = 'intermediate/all_merge_10.f'
SUBMIT = 'submit/lgbm_10.csv'

In [25]:
all_pitch = pd.read_feather(ALL_MERGE)
all_pitch.shape

(778767, 150)

### 不要な列を削除

In [26]:
all_pitch.drop(
    columns=[
        '年度', 
        '試合ID', 
        'ホームチームID', 'アウェイチームID', 
        '投手ID', '投手チームID', 
        '打者ID', '打者チームID', 
        'プレイ前走者状況', 
        '捕手ID', 
        'opening_date', 'game_date',
        'start_time', 'game_time', 'elapsed_time'
    ], inplace=True)

### Rename

In [27]:
all_pitch.rename(columns={
    'データ内連番': 'No',
    '投球位置区域': 'pitch_area',
    '試合内連番': 'No_in_game',
    '試合内投球数': 'pitch_cnt_in_game',
    'イニング': 'inning',
    'イニング内打席数': 'bat_cnt_in_inning',
    '打席内投球数': 'pitch_cnt_in_bat',
    '投手登板順': 'pitch_order',
    '投手試合内対戦打者数': 'player_cnt_in_game',
    '投手試合内投球数': 'pitcher_cnt_in_game',
    '投手イニング内投球数': 'pitcher_cnt_in_inning',
    '打者打順': 'bat_order',
    '打者試合内打席数': 'bat_cnt_in_game',
    'プレイ前ホームチーム得点数': 'home_point',
    'プレイ前アウェイチーム得点数': 'away_point',
    'プレイ前アウト数': 'out_cnt',
    'プレイ前ボール数': 'ball_cnt',
    'プレイ前ストライク数': 'strike_cnt',
}, inplace=True)

In [28]:
print(all_pitch.shape)
column_cnt = len(all_pitch.columns)
all_pitch.head()

(778767, 135)


No  ball  pitch_area  No_in_game  pitch_cnt_in_game  inning  \
0   0   0.0         0.0           1                  1       1   
1   1   0.0         8.0           2                  2       1   
2   2   0.0         5.0           3                  3       1   
3   3   0.0        12.0           4                  4       1   
4   4   0.0         8.0           5                  5       1   

   bat_cnt_in_inning  pitch_cnt_in_bat  pitch_order  player_cnt_in_game  \
0                  1                 1            1                   1   
1                  1                 2            1                   1   
2                  1                 3            1                   1   
3                  2                 1            1                   2   
4                  2                 2            1                   2   

   pitcher_cnt_in_game  pitcher_cnt_in_inning  bat_order  bat_cnt_in_game  \
0                    1                      1          1                1   
1                    2                      2          1                1   
2                    3                      3          1                1   
3                    4                      4          2                1   
4                    5                      5          2                1   

   home_point  away_point  out_cnt  ball_cnt  strike_cnt  first  second  \
0           0           0        0         0           0      0       0   
1           0           0        0         0           1      0       0   
2           0           0        0         0           2      0       0   
3           0           0        1         0           0      0       0   
4           0           0        1         0           1      0       0   

   third  base_cnt  pitch_LR  bat_LR  top_bot  role  pos_pit  \
0      0         0       0.0     1.0        1     1        0   
1      0         0       0.0     1.0        1     1        0   
2      0         0       0.0     1.0        1     1        0   
3      0         0       0.0     0.0        1     1        0   
4      0         0       0.0     0.0        1     1        0   

   date_from_opening  elapsed_min  nighter  ce-pa  league  home  point_diff  \
0                  0     0.000000        1      0       1     1           0   
1                  0     0.200000        1      0       1     1           0   
2                  0     0.383333        1      0       1     1           0   
3                  0     1.016667        1      0       1     1           0   
4                  0     1.200000        1      0       1     1           0   

   runner_23  bant  squize  cleanup  pinch  fourball  savepoint  ball_cnt_0-0  \
0          0     0       0        1      0         0          0             1   
1          0     0       0        1      0         0          0             0   
2          0     0       0        1      0         0          0             0   
3          0     0       0        1      0         0          0             1   
4          0     0       0        1      0         0          0             0   

   ball_cnt_0-1  ball_cnt_0-2  ball_cnt_0-3  ball_cnt_1-0  ball_cnt_1-1  \
0             0             0             0             0             0   
1             0             0             0             1             0   
2             0             0             0             0             0   
3             0             0             0             0             0   
4             0             0             0             1             0   

   ball_cnt_1-2  ball_cnt_1-3  ...  firm_bat  height_bat  weight_bat  \
0             0             0  ...         0         176          74   
1             0             0  ...         0         176          74   
2             0             0  ...         0         176          74   
3             0             0  ...         0         180          74   
4             0             0  ...         0         180          74   

   draft_o

### train
- 行数: 257117

In [29]:
train = all_pitch.dropna(subset=['pitch_area'])
train.shape

(257117, 135)

### test
- 行数: 521650

In [30]:
test = all_pitch[all_pitch['pitch_area'].isnull()]
test.shape

(521650, 135)

In [31]:
del all_pitch
gc.collect()

66

In [32]:
train_d = train.drop([
    'No', 
    'pitch_area', 
    'ball'
], axis=1)

In [33]:
test_d = test.drop([
    'No', 
    'pitch_area', 
    'ball'
], axis=1)

## Dataset作成

In [34]:
lgb_train = lgb.Dataset(train_d, train['ball'])

### LGBM学習
- multiclass : クラス数=8
- tuning結果のパラメータを使用
- 'learning_rate' : 0.1

In [35]:
%%time
lgb_param = {
        'objective' : 'multiclass',
        'boosting_type': 'gbdt',
        'metric' : 'multi_logloss',
        'num_class' : 8,
        'seed' : 0,
        'learning_rate' : 0.1,
        'lambda_l1': 6.9923570049658075, 
        'lambda_l2': 0.002378623984798833, 
        'num_leaves': 18, 
        'feature_fraction': 0.45199999999999996, 
        'bagging_fraction': 0.9799724836460725, 
        'bagging_freq': 4, 
        'min_child_samples': 20
    }

# 学習
cv_results = lgb.cv(lgb_param, lgb_train,
                    num_boost_round=15000,
                    early_stopping_rounds=100,
                    verbose_eval=100,
                    nfold=4)

num_boost_round = len(cv_results['multi_logloss-mean'])
print('Best num_boost_round:', num_boost_round)
best_cv_score = cv_results['multi_logloss-mean'][-1]
print('Best CV score:', best_cv_score)

[100]	cv_agg's multi_logloss: 1.15398 + 0.00199995
[200]	cv_agg's multi_logloss: 1.13238 + 0.00213603
[300]	cv_agg's multi_logloss: 1.12507 + 0.00231931
[400]	cv_agg's multi_logloss: 1.12079 + 0.00230933
[500]	cv_agg's multi_logloss: 1.11782 + 0.00223428
[600]	cv_agg's multi_logloss: 1.11553 + 0.00231126
[700]	cv_agg's multi_logloss: 1.11396 + 0.0022814
[800]	cv_agg's multi_logloss: 1.11273 + 0.0022291
[900]	cv_agg's multi_logloss: 1.11188 + 0.00234481
[1000]	cv_agg's multi_logloss: 1.11156 + 0.00240281
[1100]	cv_agg's multi_logloss: 1.1113 + 0.00247601
[1200]	cv_agg's multi_logloss: 1.1113 + 0.00263492
Best num_boost_round: 1112
Best CV score: 1.111265000190272
CPU times: user 1h 15min 40s, sys: 15.3 s, total: 1h 15min 56s
Wall time: 9min 30s


In [36]:
best_iter = int(num_boost_round * 1.1)

### 全体で再学習

In [37]:
%%time
lgb_model = lgb.train(lgb_param, lgb_train, num_boost_round=best_iter)

CPU times: user 23min 34s, sys: 5 s, total: 23min 39s
Wall time: 2min 57s


### Feature Importance

In [38]:
fi = lgb_model.feature_importance()
fn = lgb_model.feature_name()
df_feature_importance = pd.DataFrame({'feat_name':fn, 'feat_imp':fi})
df_feature_importance.sort_values('feat_imp', inplace=True)
df_feature_importance.tail(30)

feat_name  feat_imp
116            age_dif_p-b      2021
128            age_dif_b-c      2021
14                ball_cnt      2070
84              weight_bat      2202
6       player_cnt_in_game      2240
83              height_bat      2246
95         salary_year_bat      2280
60               curve_pit      2301
125            bmi_dif_p-c      2308
118  salary_x_year_dif_p-b      2330
63                fork_pit      2333
114         salary_dif_p-b      2396
126         salary_dif_b-c      2435
130  salary_x_year_dif_b-c      2519
31              point_diff      2589
89        bat_game_cnt_bat      2613
97                 bmi_bat      2791
61              slider_pit      2845
4         pitch_cnt_in_bat      2888
117    salary_year_dif_p-b      2973
88          batter_cnt_bat      3128
129    salary_year_dif_b-c      3138
0               No_in_game      3452
1        pitch_cnt_in_game      3462
131            bmi_dif_b-c      3520
119            bmi_dif_p-b      3629
8    pitcher_cnt_in_inning      3799
26             elapsed_min      4079
7      pitcher_cnt_in_game      4572
25       date_from_opening      6329

In [39]:
%%time
predict = lgb_model.predict(test_d, num_iteration = lgb_model.best_iteration)

CPU times: user 6min 33s, sys: 464 ms, total: 6min 34s
Wall time: 50.2 s


## 結果出力

In [40]:
submit = pd.DataFrame(predict)
submit.reset_index(inplace=True)
print(submit.shape)
submit.head()

(521650, 9)


index         0         1         2         3         4         5  \
0      0  0.407020  0.030354  0.114592  0.288198  0.002371  0.000026   
1      1  0.368404  0.077768  0.147787  0.307280  0.004533  0.000025   
2      2  0.286096  0.090858  0.174272  0.226682  0.089308  0.000115   
3      3  0.317413  0.093768  0.206692  0.268060  0.017942  0.000055   
4      4  0.197394  0.045490  0.306195  0.308129  0.013340  0.000050   

          6         7  
0  0.000015  0.157424  
1  0.000011  0.094191  
2  0.000025  0.132643  
3  0.000016  0.096055  
4  0.000017  0.129383

### 2017の平均
- ストレート           0.468254
- カーブ             0.074725
- スライダー           0.185806
- シュート            0.070633
- フォーク            0.083013
- チェンジアップ         0.054217
- シンカー            0.011361
- カットボール          0.051992

In [41]:
submit.mean()

index    260824.500000
0             0.521480
1             0.076928
2             0.181456
3             0.053469
4             0.071135
5             0.042502
6             0.007251
7             0.045779
dtype: float64

In [42]:
submit.to_csv(SUBMIT, header=False, index=False)

### SIGNATE

In [46]:
'signate submit --competition-id=275 ~/work/baseball/{} --note feat={}_cv={}'.format(SUBMIT, column_cnt, best_cv_score)

'signate submit --competition-id=275 ~/work/baseball/submit/lgbm_10.csv --note feat=135_cv=1.111265000190272'

### GCS

In [47]:
'gsutil cp ~/work/baseball/{} gs://yasunobu_baseball/submit/'.format(SUBMIT)

'gsutil cp ~/work/baseball/submit/lgbm_10.csv gs://yasunobu_baseball/submit/'

### 評価結果
#### 10 (135) ... 5相当に戻した
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1112
- Best CV score: 1.111265000190272
- Wall time: 9min 30s
- 評価結果  : 1.4008712

#### 9 (136) ... 5相当に戻した=>NG
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1106
- Best CV score: 1.1113177029272225
- Wall time: 25min 44s (local)
- 評価結果  : 1.5133521

#### 8 (157)
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1100
- Best CV score: 1.1093728644236824
- Wall time: 10min 4s
- 評価結果  : 1.4925832

#### 7 (235)
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1204
- Best CV score: 1.108590082437519
- Wall time: 12min 56s
- 評価結果  : 1.4946937

#### 6 (207)
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1135
- Best CV score: 1.1086837935965734
- Wall time: 11min 54s
- 評価結果  : 1.3999194

#### 5-tune (参考)
- 'num_leaves' : 18,
- 'learning_rate' : 0.05,
- Best num_boost_round: 2395
- Best CV score: 1.1103399384119979
- Wall time: 18min 29s (n1-standard-8)

#### 5-tune (135)
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- CV score: 1.1112402566616455
- 評価結果  : 1.3991780
- CV Wall time: 27min 48s (local) -> 18min 24s (n1-standard-4) -> 9min 3s (n1-standard-8)

#### 5 (135)
- 'num_leaves' : 31,
- 'learning_rate' : 0.1,
- CV score: 1.1161825304353852
- 評価結果  : 1.4252286

#### 4 (128)
- CV score: 1.1158111693845418
- 評価結果  : 1.4299265

#### 3 (97)
- CV score: 1.1172290313076036
- 評価結果  : 1.4335806